---
layout: default
title: Single cell RNA-seq analysis 
---

# Single cell RNA-seq analysis  

This workshop will introduce the best practice bioinformatics methods for processing and analyses of single cell RNA-seq data via a series of lectures and computer practicals. The total course duration is 5 days, with 4 course days followed by a Bring Your Own Data day where you can apply the skills learned on a dataset of your chosing.

**When:** 12-16 February 2024  (9.00-17.00)

**Location:** TrippelrummetLinks to an external site., Entrance C11, BMC, Husargatan 3, Uppsala 75237, Sweden


### SYLLABUS

* Overview of the current scRNAseq technologies
* Basic overview of pipelines for processing raw reads into expression values
* Quality control and normalization
* Dimensionality reduction techniques
* Data integration and batch correction
* Differential gene expression
* Clustering techniques
* Celltype prediction
* Trajectory inference analysis
* Comparison of different Seurat, Bioconductor and Scanpy toolkits

<br/>

### REGISTRATION

**Open:** 2023-11-06

**Close:** 2024-01-17

**Application form:** [application](https://nbis.typeform.com/to/VYqpkdkhLinks)

Course fee: This workshop has a fee of 3000 SEK and will be invoiced to the selected participants (Please note that NBIS cannot invoice individuals). Applications without complete invoice information will not be considered. Course fees cover all coffee breaks, all lunches and 1 course dinner.

This is a national course. The course is open for PhD students, postdocs, group leaders and core facility staff within all Swedish universities. Even if we do accept application from other countries, we give priority to applicants from Swedish universities prior to applicants from industry and academics from other countries.

Please note that NBIS training events do not provide any formal university credits. The training content is estimated to correspond to a certain number of credits, however the estimated credits are just guidelines. If formal credits are crucial, the student needs to confer with the home department before submitting a course application in order to establish whether the course is valid for formal credits or not.

Practical exercises will be performed using either R or Python, so we only accept students that fulfill the entry requirements below.

**Entry requirements:**

* You are familiar with basic UNIX command line (bash)
* You are familiar with either R and/or Python
* You already understand the basics of NGS technologies
* You are familiar with analyses of bulk RNA-sequencing data
* You have full access to your computer (admin permissions). Instructions on installation will be sent by email to accepted participants.

Due to limited space the course can accommodate maximum of 25 participants. If we receive more applications, participants will be selected based on several criteria. Selection criteria include correct entry requirements, motivation to attend the course as well as gender and geographical balance.


<br/>


### CONTACT

Åsa Björklund, Susanne Reinsbach, Roy Francis.

edu.sc[at]nbis.se 



<br/>



##### <img border="0" src="https://www.svgrepo.com/show/158264/schedule.svg" width="40" height="40"> Schedule
***

Course schedule can be found here: [schedule](schedule.md) (Will be updated, for now is last years schedule)

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/6672/exercise.svg" width="40" height="40"> Exercises
***

All exercises can be found at [Exercises](exercises.md). (TBA)

For working on Uppmax: to use the allocations we have for the course, please look [here](login.md).

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/19652/maths-class-materials-cross-of-a-pencil-and-a-ruler.svg" width="40" height="40"> Precourse material
***

Please read carefully the [Precourse material](precourse.md) before the course start.

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/38706/group-of-people.svg" width="40" height="40"> Course leaders
***

If you have any questions regarding the course, please email to edu.sc@nbis.se.

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/19262/link.svg" width="40" height="40"> Useful links
***

* [Single cell RNA-seq course at from Hemberg lab](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html)
* [Single cell RNA-seq course in Python](https://chanzuckerberg.github.io/scRNA-python-workshop/intro/about)
* [Single cell RNA-seq course at Broad](https://broadinstitute.github.io/2019_scWorkshop/)
* [Repository listing many scRNA-seq tools](https://github.com/seandavi/awesome-single-cell)
* [SingleCellExperiment objects for many datasets](https://hemberg-lab.github.io/scRNA.seq.datasets/)
* [Conquer datasets - many different datasets based on a salmon pipeline](http://imlspenticton.uzh.ch:3838/conquer/)
* [The Human Cell Atlas project](https://www.humancellatlas.org/)
* [The github repository for this course. Also contains the R markdown scripts that was used to generate all the exercises](https://github.com/NBISweden/workshop-scRNAseq)
* [Bitbucket repository for QC report scripts](https://bitbucket.org/asbj/qc-summary_scrnaseq)
* [Bitbucket repository for the NBIS scRNAseq pipeline](https://bitbucket.org/scilifelab-lts/lts-workflows-sm-scrnaseq)

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/83468/navigation-history-interface-symbol-of-a-clock-with-an-arrow.svg" width="40" height="40"> Past courses
***

* [Stockholm January 27-30, 2020](https://nbisweden.github.io/workshop-archive/workshop-scRNAseq/2020-01-27/)
* [Stockholm February 4-6, 2019](https://nbisweden.github.io/workshop-archive/workshop-scRNAseq/2019-02-04/)
* [Stockholm May 21-23, 2018](https://nbisweden.github.io/workshop-archive/workshop-scRNAseq/2018-05-21/)
* [Uppsala October 2-4, 2017](https://scilifelab.github.io/courses/scrnaseq/1710/)

<br/>

<div style="text-align: right; font-size: 5px"> Icons are provided from [www.svgrepo.com](www.svgrepo.com) </div>
