---
title: "Scater/Scran:: Spatial Transcriptomics"
#CSS_ALL:
editor_options:
  chunk_output_type: console
---

#CHUNK_OPT:

#ST_TITLE:


#ST_ALL1:

#ST_TITLE1:

```{r}
webpath <- "https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Posterior/"
PATH <- "./data/visium/Posterior"
if(!dir.exists(PATH)){dir.create(PATH,recursive = T)}
file_list <- c("V1_Mouse_Brain_Sagittal_Posterior_filtered_feature_bc_matrix.tar.gz",
               "V1_Mouse_Brain_Sagittal_Posterior_spatial.tar.gz")
for(i in file_list){  download.file( url = paste0(webpath,i) , destfile = paste0("./data/raw/",i))
  system(paste0("tar xvzf ./data/raw/",i))}

webpath <- "https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior/"
PATH <- "./data/visium/Anterior"
if(!dir.exists(PATH)){dir.create(PATH,recursive = T)}
file_list <- c("V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.tar.gz",
               "V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz")
for(i in file_list){  download.file( url = paste0(webpath,i) , destfile = paste0("./data/raw/",i))
  system(paste0("tar xvzf ./data/raw/",i))}
```


```{r packages}
#BiocManager::install('DropletUtils',update = F)
devtools::install_github("RachelQueen1/Spaniel", ref = "Development" ,upgrade = F,dependencies = F)

library(Spaniel)
library(biomaRt)

suppressPackageStartupMessages(require(Matrix))
suppressPackageStartupMessages(require(dplyr))
suppressPackageStartupMessages(require(scran))
suppressPackageStartupMessages(require(SingleR))
suppressPackageStartupMessages(require(scater))
suppressPackageStartupMessages(require(ggplot2))
suppressPackageStartupMessages(require(patchwork))
suppressPackageStartupMessages(require(cowplot))
```


#ST_TITLE2:

We can first load and merge the objects into one SCE object.

```{r}
sce.a <- Spaniel::createVisiumSCE(tenXDir="data/visium/Anterior",
                            resolution="Low")
sce.p <- Spaniel::createVisiumSCE(tenXDir="data/visium/Posterior",
                            resolution="Low")

sce <- cbind(sce.a , sce.p)
sce$Sample <- sub( ".*[/]" , "" ,  sub( "/filtered_feature_bc_matrix" , "" , sce$Sample ) )

lll <- list(sce.a , sce.p)
lll <- lapply( lll, function(x) x@metadata )
names(lll) <- c("Anterior","Posterior")
sce@metadata <- lll
```

We can further convert the gene ensembl IDs to gene names.

```{r}
mart <- biomaRt::useMart(biomart="ENSEMBL_MART_ENSEMBL", dataset="mmusculus_gene_ensembl")
annot <- biomaRt::getBM(attributes = c("ensembl_gene_id", "external_gene_name","gene_biotype"), mart = mart, useCache = F)

gene_names <- as.character ( annot[match(rownames(sce),annot[,"ensembl_gene_id"]),"external_gene_name"] )
gene_names[is.na(gene_names) ] <- ""

sce <- sce[gene_names != "",]
rownames(sce) <- gene_names[gene_names != ""]
dim(sce)
```


#ST_TITLE3:

#ST_ALL3:

Now the counts and feature counts are calculated on the Spatial assay, so they are named  "nCount_Spatial" and "nFeature_Spatial".

```{r, fig.height=10}
# Mitochondrial genes
mito_genes <- rownames(sce)[grep("^mt-",rownames(sce))]

# Ribosomal genes
ribo_genes <- rownames(sce)[grep("^Rp[sl]",rownames(sce))]

# Hemoglobin genes - includes all genes starting with HB except HBP.
hb_genes <- rownames(sce)[grep("^Hb[^(p)]",rownames(sce))]

sce <- addPerCellQC(sce, flatten = T, subsets = list(mt=mito_genes, hb=hb_genes, ribo=ribo_genes))

head(colData(sce))

plot_grid(plotColData(sce,y = "detected",x = "Sample",colour_by = "Sample"),
          plotColData(sce,y = "total",x = "Sample",colour_by = "Sample"),
          plotColData(sce,y = "subsets_mt_percent",x = "Sample",colour_by = "Sample"),
          plotColData(sce,y = "subsets_ribo_percent",x = "Sample",colour_by = "Sample"),
          plotColData(sce,y = "subsets_hb_percent",x = "Sample",colour_by = "Sample"),ncol = 3)

```

#ST_ALL3.1:

```{r, fig.height=12}
samples <- c("Anterior", "Posterior")
to_plot <- c("detected","total","subsets_mt_percent","subsets_ribo_percent","subsets_hb_percent")

plist <- list()
n=1
for(j in to_plot){
  for(i in samples){
    temp <- sce[,sce$Sample == i]
    temp@metadata <- temp@metadata[[i]]
    plist[[n]]<-spanielPlot(object = temp,
        plotType = "Cluster",
        clusterRes=j,customTitle = j,
        techType = "Visium",
        ptSizeMax = 1,ptSizeMin = .1)
    n<-n+1
  }
}

plot_grid(ncol=2, plotlist = plist)
```


#ST_ALL3.2:

#ST_TITLE4:

#ST_ALL4:


```{r}
sce <- sce[ , sce$detected > 500 &
              sce$subsets_mt_percent < 25 &
              sce$subsets_hb_percent < 20]
dim(sce)
```

And replot onto tissue section:

```{r, fig.height=10}
samples <- c("Anterior", "Posterior")
to_plot <- c("detected","total","subsets_mt_percent","subsets_mt_percent","subsets_hb_percent")

plist <- list()
n=1
for(j in to_plot){
  for(i in samples){
    temp <- sce[,sce$Sample == i]
    temp@metadata <- temp@metadata[[i]]
    plist[[n]]<-spanielPlot(object = temp,
        plotType = "Cluster",
        clusterRes=j,customTitle = j,
        techType = "Visium",
        ptSizeMax = 1,ptSizeMin = .1)
    n<-n+1
  }
}

plot_grid(ncol=2, plotlist = plist)
```

#ST_TITLE5:

```{r, fig.height=8, fig.width=6}
C = counts(sce)
C@x = C@x / rep.int(colSums(C), diff(C@p))
most_expressed <- order(Matrix::rowSums( C ),decreasing = T)[20:1]
boxplot( as.matrix(t(C[most_expressed,])),cex=.1, las=1, xlab="% total count per cell",col=scales::hue_pal()(20)[20:1],horizontal=TRUE)
rm(C)
```

#ST_ALL5:

#ST_TITLE6:

```{r}
dim(sce)

# Filter Bl1
sce <- sce[!grepl("Bc1", rownames(sce)), ]

# Filter Mitocondrial
sce <- sce[!grepl("^mt-", rownames(sce)), ]

# Filter Hemoglobin gene (optional if that is a problem on your data)
sce <- sce[!grepl("^Hb.*-", rownames(sce)), ]

dim(sce)
```

#ST_TITLE7:

```{r}
sce <- computeSumFactors(sce, sizes=c(20, 40, 60, 80))
sce <- logNormCounts(sce)
```


#ST_ALL7:

```{r}
samples <- c("Anterior", "Posterior")
to_plot <- c("Hpca", "Ttr")

plist <- list()
n=1
for(j in to_plot){
  for(i in samples){
    temp <- sce[,sce$Sample == i]
    temp@metadata <- temp@metadata[[i]]
    plist[[n]]<-spanielPlot(object = temp,
        plotType = "Gene",
        gene=j,
        customTitle = j,
        techType = "Visium",
        ptSizeMax = 1,ptSizeMin = .1)
    n<-n+1
  }
}

plot_grid(ncol=2, plotlist = plist)
```


#ST_ALL7.1:

But make sure you run it on the `SCT` assay.

```{r}
var.out <- modelGeneVar(sce, method="loess")
hvgs = getTopHVGs(var.out, n=2000)
sce <- runPCA(sce, exprs_values = "logcounts",
              subset_row=hvgs,
              ncomponents = 50,
              ntop = 100,
              scale = T)
g <- buildSNNGraph(sce,k=5,use.dimred="PCA")
sce$louvain_SNNk5 <- factor( igraph::cluster_louvain(g)$membership )
sce <- runUMAP(sce,dimred = "PCA", n_dimred = 50,   ncomponents = 2,min_dist=0.1,spread=.3,
               metric="correlation",name = "UMAP_on_PCA")
```

#ST_ALL7.2:

```{r}
samples <- c("Anterior", "Posterior")
to_plot <- c("louvain_SNNk5")

plist <- list()
n=1
for(j in to_plot){
  for(i in samples){
    temp <- sce[,sce$Sample == i]
    temp@metadata <- temp@metadata[[i]]
    plist[[n]]<-spanielPlot(object = temp,
        plotType = "Cluster",clusterRes = j,
        customTitle = j,
        techType = "Visium",
        ptSizeMax = 1,ptSizeMin = .1)
    n<-n+1
  }
}

plist[[3]] <- plotReducedDim(sce, dimred = "UMAP_on_PCA", colour_by = "louvain_SNNk5")
plist[[4]] <- plotReducedDim(sce, dimred = "UMAP_on_PCA", colour_by = "Sample")

plot_grid(ncol=2, plotlist = plist)

```


#ST_TITLE8:

#ST_ALL8:

We will do a similar integration as in the Data Integration lab.

```{r}
mnn_out <- batchelor::fastMNN(sce, subset.row = hvgs,
                              batch = factor(sce$Sample),
                              k = 20, d = 50)

reducedDim(sce, "MNN") <- reducedDim(mnn_out,"corrected")
rm(mnn_out)
gc()
```


#ST_ALL8.1:

```{r}
g <- buildSNNGraph(sce,k=5,use.dimred="MNN")
sce$louvain_SNNk5 <- factor( igraph::cluster_louvain(g)$membership )
sce <- runUMAP(sce,dimred = "MNN", n_dimred = 50,   ncomponents = 2,min_dist=0.1,spread=.3,
               metric="correlation",name = "UMAP_on_MNN")
```

```{r}
samples <- c("Anterior", "Posterior")
to_plot <- c("louvain_SNNk5")

plist <- list()
n=1
for(j in to_plot){
  for(i in samples){
    temp <- sce[,sce$Sample == i]
    temp@metadata <- temp@metadata[[i]]
    plist[[n]]<-spanielPlot(object = temp,
        plotType = "Cluster",clusterRes = j,
        customTitle = j,
        techType = "Visium",
        ptSizeMax = 1,ptSizeMin = .1)
    n<-n+1
  }
}

plist[[3]] <- plotReducedDim(sce, dimred = "UMAP_on_MNN", colour_by = "louvain_SNNk5")
plist[[4]] <- plotReducedDim(sce, dimred = "UMAP_on_MNN", colour_by = "Sample")

plot_grid(ncol=2, plotlist = plist)
```

#ST_ALL8.2:

#ST_TITLE9:

#ST_ALL9:

```{r, fig.height=12}
# differential expression between cluster 4 and cluster 6
cell_selection <- sce[ , sce$louvain_SNNk5 %in% c(4,6) ]
cell_selection$louvain_SNNk5 <- factor(cell_selection$louvain_SNNk5)

markers_genes <- scran::findMarkers( x = cell_selection,
                              groups = cell_selection$louvain_SNNk5,
                              lfc=.25,
                              pval.type = "all",
                              direction = "up")

#List of dataFrames with the results for each cluster
top5_cell_selection <- lapply( names(markers_genes), function(x) { temp <- markers_genes[[x]][1:5, 1:2] ; temp$gene <- rownames(markers_genes[[x]])[1:5] ; temp$cluster <- x ; return(temp) } )
top5_cell_selection <- as_tibble(do.call(rbind, top5_cell_selection))
top5_cell_selection


# plot top markers
samples <- c("Anterior", "Posterior")
to_plot <- top5_cell_selection$gene[1:5]

plist <- list()
n=1
for(j in to_plot){
  for(i in samples){
    temp <- sce[,sce$Sample == i]
    temp@metadata <- temp@metadata[[i]]
    plist[[n]]<-spanielPlot(object = temp,
        plotType = "Gene",
        gene=j,
        customTitle = j,
        techType = "Visium",
        ptSizeMax = 1,ptSizeMin = .1)
    n<-n+1
  }
}
plot_grid(ncol=2, plotlist = plist)
```


#ST_TITLE10:

#ST_ALL10:


First dowload the seurat data from: https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1 to folder `data/spatial/` with command:

```{r}
webpath <- "https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1"
PATH <- "./data/spatial/allen_cortex.rds"
if(!file.exists(PATH)){  
  dir.create("./data/spatial/",recursive = T)
  options(timeout=10000)
  download.file( url = webpath , destfile = PATH)
  options(timeout=60)}
```

#ST_ALL10.1:

```{r subset_sc}
allen_reference <- readRDS("data/spatial/allen_cortex.rds")
allen_reference_sce <- Seurat::as.SingleCellExperiment(allen_reference)

# check number of cells per subclass
allen_reference_sce$subclass <- sub("/","_",sub(" ","_",allen_reference_sce$subclass))
table(allen_reference_sce$subclass)

# select 20 cells per subclass, fist set subclass ass active.ident
subset_cells <- lapply(unique(allen_reference_sce$subclass),function(x){
  if( sum(allen_reference_sce$subclass == x) > 20 ){
    temp <- sample(colnames(allen_reference_sce)[allen_reference_sce$subclass == x],size = 20)
  } else {
    temp <- colnames(allen_reference_sce)[allen_reference_sce$subclass == x]
  } })
allen_reference_sce <- allen_reference_sce[,unlist(subset_cells)]

# check again number of cells per subclass
table(allen_reference_sce$subclass)
```

Then run normalization and dimensionality reduction.

```{r}
allen_reference_sce <- computeSumFactors(allen_reference_sce, sizes=c(20, 40, 60, 80))
allen_reference_sce <- logNormCounts(allen_reference_sce)
allen.var.out <- modelGeneVar(allen_reference_sce, method="loess")
allen.hvgs = getTopHVGs(allen.var.out, n=2000)
```


#ST_ALL10.1b:


#ST_ALL10.2:

Here, will use SingleR for prediciting which cell types are present in the dataset.

We can first select the anterior part as an example (to speed up predictions).

```{r}
sce.anterior <- sce[,sce$Sample == "Anterior"]
sce.anterior@metadata <- sce.anterior@metadata[["Anterior"]]
```

Next, we select the highly variable genes that are present in both datasets.

```{r}
#Find common highly variable genes
common_hvgs <- allen.hvgs[allen.hvgs %in% hvgs]

#Predict cell classes
pred.grun <- SingleR(test=sce.anterior[common_hvgs,],
                     ref=allen_reference_sce[common_hvgs,],
                     labels=allen_reference_sce$subclass)

#Transfer the classes to the SCE object
sce.anterior$cell_prediction <- pred.grun$labels
sce.anterior@colData <- cbind(sce.anterior@colData,
                              as.data.frame.matrix(table(list(1:ncol(sce.anterior),sce.anterior$cell_prediction))))
```

Then we can plot the predicted cell populations back to tissue.

```{r}
#Plot cell predictions
spanielPlot(object = sce.anterior,
        plotType = "Cluster",
        clusterRes = "cell_prediction",
        customTitle = "cell_prediction",
        techType = "Visium",
        ptSizeMax = 1,ptSizeMin = .1)


plist <- list()
n=1
for(i in c("L2_3_IT","L4","L5_IT","L6_IT")){
  plist[[n]] <- spanielPlot(object = sce.anterior,
        plotType = "Cluster",
        clusterRes = i,
        customTitle = i,
        techType = "Visium",ptSize = .3,
        ptSizeMax = 1,ptSizeMin = .1)
  n <- n+1
}
plot_grid(ncol=2, plotlist = plist)
```


#ST_ALL10.4:


We can also plot the gene expression and add filters together, too:

```{r}
spanielPlot(object = sce.anterior,
        plotType = "Gene",
        gene="Wfs1",
        showFilter = sce.anterior$L4 ,
        customTitle = "",
        techType = "Visium",
        ptSize=0,ptSizeMin = -.3,ptSizeMax = 1)

```


### Session info

```{r}
sessionInfo()
```
