---
#CSS_ALL:
editor_options:
  chunk_output_type: console
---

#CHUNK_OPT:

#CT_TITLE:

#CT_ALL1:

#CT_SCRAN1:

#CT_ALL2:

```{r}
suppressPackageStartupMessages({
  library(scater)
  library(scran)
  library(dplyr)
  library(cowplot)
  library(ggplot2)
  library(pheatmap)
  library(rafalib)
  library(scPred)
  library(scmap)
})
```



```{r}
#load the data and select 'ctrl_13` sample
alldata <- readRDS("data/results/covid_qc_dr_int_cl.rds")
ctrl.sce <- alldata[, alldata@colData$sample == 'ctrl.13']

# remove all old dimensionality reductions as they will mess up the analysis further down
reducedDims(ctrl.sce) <- NULL
```

#CT_ALL3:


```{r}
reference <- scPred::pbmc_1

reference
```

Convert to a SCE object.

```{r}
ref.sce = Seurat::as.SingleCellExperiment(reference)
```


#CT_SCRAN3:

```{r}
#Normalize
ref.sce <- computeSumFactors(ref.sce)
ref.sce <- logNormCounts(ref.sce)

#Variable genes
var.out <- modelGeneVar(ref.sce, method="loess")
hvg.ref <- getTopHVGs(var.out, n=1000)  

# Dim reduction
ref.sce <- runPCA(ref.sce, exprs_values = "logcounts",  scale = T,
              ncomponents = 30, subset_row = hvg.ref)
ref.sce <- runUMAP(ref.sce, dimred = "PCA")

```

```{r, fig.width=5}
plotReducedDim(ref.sce,dimred = "UMAP",colour_by = "cell_type")

```


#CT_ALL4:

```{r}
#Normalize
ctrl.sce <- computeSumFactors(ctrl.sce)
ctrl.sce <- logNormCounts(ctrl.sce)

#Variable genes
var.out <- modelGeneVar(ctrl.sce, method="loess")
hvg.ctrl <- getTopHVGs(var.out, n=1000)  

# Dim reduction
ctrl.sce <- runPCA(ctrl.sce, exprs_values = "logcounts",  scale = T,
              ncomponents = 30, subset_row = hvg.ctrl)
ctrl.sce <- runUMAP(ctrl.sce, dimred = "PCA")

```

```{r, fig.width=5}
plotReducedDim(ctrl.sce,dimred = "UMAP",colour_by = "louvain_SNNk15")

```

#CT_SCRAN5:

#CT_SCRAN5.1:


```{r}

# add in slot cell_type1
ref.sce@colData$cell_type1 = ref.sce@colData$cell_type
# create a rowData slot with feature_symbol
rd = data.frame(feature_symbol=rownames(ref.sce))
rownames(rd) = rownames(ref.sce)
rowData(ref.sce) = rd




# same for the ctrl dataset
# create a rowData slot with feature_symbol
rd = data.frame(feature_symbol=rownames(ctrl.sce))
rownames(rd) = rownames(ctrl.sce)
rowData(ctrl.sce) = rd


# select features
counts(ctrl.sce) <- as.matrix(counts(ctrl.sce))
logcounts(ctrl.sce) <- as.matrix(logcounts(ctrl.sce))
ctrl.sce <- selectFeatures(ctrl.sce, suppress_plot = TRUE)

counts(ref.sce) <- as.matrix(counts(ref.sce))
logcounts(ref.sce) <- as.matrix(logcounts(ref.sce))
ref.sce <- selectFeatures(ref.sce, suppress_plot = TRUE)
```

#CT_SCRAN5.2:

```{r}
ref.sce <- indexCluster(ref.sce)

```

#CT_SCRAN5.3:

```{r}
project_cluster <- scmapCluster(
  projection = ctrl.sce,
  index_list = list(
    ref = metadata(ref.sce)$scmap_cluster_index
  )
)

# projected labels
table(project_cluster$scmap_cluster_labs)

```

#CT_SCRAN5.4:

```{r}
# add in predictions
ctrl.sce@colData$scmap_cluster <- project_cluster$scmap_cluster_labs

plotReducedDim(ctrl.sce,dimred = "UMAP",colour_by = "scmap_cluster")

```

#CT_SCRAN5.5:

```{r}
ref.sce <- indexCell(ref.sce)
```

#CT_SCRAN5.6:

```{r}
project_cell <- scmapCell(
  projection = ctrl.sce,
  index_list = list(
    ref = metadata(ref.sce)$scmap_cell_index
  )
)


```

#CT_SCRAN5.7:

```{r}
cell_type_pred <- colData(ref.sce)$cell_type1[project_cell$ref[[1]][1,]]

table(cell_type_pred)
```


#CT_SCRAN5.8:

```{r}
# add in predictions
ctrl.sce@colData$scmap_cell <- cell_type_pred

plotReducedDim(ctrl.sce,dimred = "UMAP",colour_by = "scmap_cell")

```

Plot both:

```{r}
cowplot::plot_grid( ncol = 2,
plotReducedDim(ctrl.sce,dimred = "UMAP",colour_by = "scmap_cluster"),
plotReducedDim(ctrl.sce,dimred = "UMAP",colour_by = "scmap_cell")
)

```



#CT_SEURAT5:

scPred works with Seurat objects, so we will convert both objects to seurat objects. You may see a lot of warnings about renaming things, but as long as you do not see an Error, you should be fine.

```{r}
suppressPackageStartupMessages(library(Seurat))

reference <- Seurat::as.Seurat(ref.sce)
ctrl <- Seurat::as.Seurat(ctrl.sce)
```

The loadings matrix is lost when converted to Seurat object, and scPred needs that information. So we need to rerun PCA with Seurat and the same hvgs.

```{r}
VariableFeatures(reference) = hvg.ref
reference <- reference %>% ScaleData(verbose=F) %>% RunPCA(verbose=F)

VariableFeatures(ctrl) = hvg.ctrl
ctrl <- ctrl %>% ScaleData(verbose=F) %>% RunPCA(verbose=F)

```



```{r}
reference <- getFeatureSpace(reference, "cell_type")

reference <- trainModel(reference)
```



#CT_SEURAT5.2:

```{r}
get_scpred(reference)
```

#CT_SEURAT5.3:

```{r}
ctrl <- scPredict(ctrl, reference)
```

```{r}
DimPlot(ctrl, group.by = "scpred_prediction", label = T, repel = T) + NoAxes()
```

#CT_SEURAT5.4:

```{r}
ggplot(ctrl@meta.data, aes(x=louvain_SNNk15, fill = scpred_prediction)) + geom_bar() + theme_classic()
```

Add the predictions into the SCE object

```{r}
ctrl.sce@colData$scpred_prediction = ctrl$scpred_prediction
```

#CT_ALL6:

#CT_SEURAT6:

```{r}
crossTab(ctrl, "scmap_cell", "scpred_prediction")
```


#CT_GSEA_TITLE:

#CT_GSEA1:

#CT_GSEA2:

```{r}
# run differential expression in our dataset, using clustering at resolution 0.3
DGE_list <- scran::findMarkers( x = alldata,
                              groups = as.character(alldata@colData$louvain_SNNk15),
                              pval.type = "all",
                              min.prop = 0)

```

```{r}
# Compute differential gene expression in reference dataset (that has cell annotation)
ref_DGE <- scran::findMarkers( x = ref.sce,
                              groups = as.character(ref.sce@colData$cell_type),
                              pval.type = "all",
                              direction = "up")


# Identify the top cell marker genes in reference dataset
# select top 50 with hihgest foldchange among top 100 signifcant genes.
ref_list <- lapply(ref_DGE, function(x){
  x$logFC <- rowSums( as.matrix( x[,grep("logFC",colnames(x))] ))
  x %>% as.data.frame() %>% filter(p.value < 0.01) %>% top_n(-100, p.value) %>% top_n(50, logFC) %>% rownames()
  }
)

unlist(lapply(ref_list, length))
```


#CT_GSEA3:

```{r}
suppressPackageStartupMessages(library(fgsea))

# run fgsea for each of the clusters in the list
res <- lapply(DGE_list, function(x){
  x$logFC <- rowSums( as.matrix( x[,grep("logFC",colnames(x))] ))
  gene_rank <- setNames(x$logFC, rownames(x))
  fgseaRes <- fgsea( pathways=ref_list, stats=gene_rank,nperm=10000)
  return(fgseaRes)
})
names(res) <- names(DGE_list)

# You can filter and resort the table based on ES, NES or pvalue
res <- lapply(res, function(x) {x[ x$pval < 0.1 , ]} )
res <- lapply(res, function(x) {x[ x$size > 2 , ]} )
res <- lapply(res, function(x) {x[order(x$NES,decreasing = T), ]} )
res
```

#CT_GSEA4:

```{r}
new.cluster.ids <- unlist(lapply(res,function(x){as.data.frame(x)[1,1]}))

alldata@colData$ref_gsea <- new.cluster.ids[as.character(alldata@colData$louvain_SNNk15)]

cowplot::plot_grid( ncol = 2,
   plotReducedDim(alldata, dimred = "UMAP",colour_by = "louvain_SNNk15"),
   plotReducedDim(alldata, dimred = "UMAP",colour_by = "ref_gsea")
)

```

#CT_GSEA5:

```{r, fig.width=10}
ctrl.sce@colData$ref_gsea = alldata@colData$ref_gsea[alldata@colData$sample == "ctrl.13"]

cowplot::plot_grid( ncol = 3,
  plotReducedDim(ctrl.sce, dimred = "UMAP",colour_by = "ref_gsea"),
  plotReducedDim(ctrl.sce, dimred = "UMAP",colour_by = "scmap_cell"),          
  plotReducedDim(ctrl.sce, dimred = "UMAP",colour_by = "scpred_prediction")
)
```

#CT_GSEA6:

```{r}
# Download gene marker list
if(!dir.exists("data/CellMarker_list/")) {
  dir.create("data/CellMarker_list")
  download.file(url="http://bio-bigdata.hrbmu.edu.cn/CellMarker/download/Human_cell_markers.txt",
                destfile = "./data/CellMarker_list/Human_cell_markers.txt")
  download.file(url="http://bio-bigdata.hrbmu.edu.cn/CellMarker/download/Mouse_cell_markers.txt",
                destfile = "./data/CellMarker_list/Mouse_cell_markers.txt")

}
```

#CT_GSEA7:

```{r}

# Load the human marker table
markers <- read.delim("data/CellMarker_list/Human_cell_markers.txt")
markers <- markers [ markers$speciesType == "Human", ]
markers <- markers [ markers$cancerType == "Normal", ]

#Filter by tissue (to reduce computational time and have tissue-specific classification)
# sort(unique(markers$tissueType))
# grep("blood",unique(markers$tissueType),value = T)
# markers <- markers [ markers$tissueType %in% c("Blood","Venous blood",
#                                                "Serum","Plasma",
#                                                "Spleen","Bone marrow","Lymph node"), ]


# remove strange characters etc.
celltype_list <- lapply( unique(markers$cellName) , function(x){
  x <- paste(markers$geneSymbol[markers$cellName == x],sep=",")
  x <- gsub("[[]|[]]| |-",",",x)
  x <- unlist(strsplit( x , split = ","))
  x <- unique(x [ ! x %in% c("","NA","family") ])
  x <- casefold(x,upper = T)
})
names(celltype_list) <- unique(markers$cellName)
# celltype_list <- lapply(celltype_list , function(x) {x[1:min(length(x),50)]} )
celltype_list <- celltype_list[ unlist(lapply(celltype_list,length)) < 100 ]
celltype_list <- celltype_list[ unlist(lapply(celltype_list,length)) > 5 ]

```

```{r}
# run fgsea for each of the clusters in the list
res <- lapply(DGE_list, function(x){
  x$logFC <- rowSums( as.matrix( x[,grep("logFC",colnames(x))] ))
  gene_rank <- setNames(x$logFC, rownames(x))
  fgseaRes <- fgsea( pathways=celltype_list, stats=gene_rank,nperm=10000)
  return(fgseaRes)
})
names(res) <- names(DGE_list)


# You can filter and resort the table based on ES, NES or pvalue
res <- lapply(res, function(x) {x[ x$pval < 0.01 , ]} )
res <- lapply(res, function(x) {x[ x$size > 5 , ]} )
res <- lapply(res, function(x) {x[order(x$NES,decreasing = T), ]} )

# show top 3 for each cluster.
lapply(res,head,3)
```

#CT_GSEA8:

```{r}
new.cluster.ids <- unlist(lapply(res,function(x){as.data.frame(x)[1,1]}))
alldata@colData$cellmarker_gsea <- new.cluster.ids[as.character(alldata@colData$louvain_SNNk15)]

cowplot::plot_grid( ncol = 2,
 plotReducedDim(alldata,dimred = "UMAP",colour_by = "cellmarker_gsea"),
 plotReducedDim(alldata,dimred = "UMAP",colour_by = "ref_gsea")                  
)

```




#CT_ALL7:

```{r}
saveRDS(ctrl.sce,"data/results/ctrl13_qc_dr_int_cl_celltype.rds")
```


#SESSION_INFO:

```{r}
sessionInfo()
```
