---
#CSS_ALL:
---

#CHUNK_OPT:

# Trajectory inference analysis: Slingshot


### Downloading dataset

```{bash, eval=F}
#Create data folder
mkdir data

#Download file from NCBI server into data folder
curl ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE72nnn/GSE72857/suppl/GSE72857%5Fumitab%2Etxt%2Egz -o data/GSE72857_umitab.txt.gz

#Decompress it
gunzip GSE72857_series_matrix.txt.gz
```


### Loading matrix into R and converting it to compressed `.rds` format.

```{r, eval=FALSE}
data <- read.delim("data/GSE72857_umitab.txt", header = T, row.names = 1)
comp_matrix <- Matrix::Matrix(as.matrix(data),sparse = T)
saveRDS(comp_matrix,"data/GSE72857_umitab.rds")
```



### Loading data

In order to speed up the computations during the exercises, we will be selecting every fith cell of the dataset.

```{r}
umi_counts <- readRDS("data/GSE72857_umitab.rds")
umi_counts <- umi_counts[ , c(T,F,F,F,F)]
dim(umi_counts)

#Define a color pallete to use
pal <- c( RColorBrewer::brewer.pal(9, "Set1"), RColorBrewer::brewer.pal(8, "Set2"))
```


### Basic processing with Scran/Scater pipeline
***

If you have been using the `scran/scater` pipeline so far. Below you can find a summary code for data processing until getting:

* A reduction where to perform the trajectory (UMAP)
* The cell clustering information

```{r, fig.height=10,fig.width=10}
suppressPackageStartupMessages({
library(scran)
library(scater)
library(igraph)
})

#Data analysis with Scran pipeline
data <- SingleCellExperiment( assays = list(counts = umi_counts) )
data <- computeSumFactors(data, sizes=c(20, 40, 60, 80))
data <- logNormCounts(data)
var.fit <- modelGeneVar(data)
top.hvgs <- getTopHVGs(var.fit, n=2000)
data <- runPCA(data, subset_row=top.hvgs)
g <- buildSNNGraph(data, k=5, use.dimred="PCA", assay.type="RNA")
data$louvain_SNNk5 <- factor( igraph::cluster_louvain(g)$membership )
data <- runUMAP(data, dimred = "PCA", n_dimred = 50,   ncomponents = 2, spread=2, min_dist = .1, n_neighbors=10, metric="cosine")

#Plot the clusters
plotReducedDim(data, dimred = "UMAP",colour_by = "louvain_SNNk5")

#Save the objects as separate matrices for input in slingshot
dimred <- reducedDim(data, type = "UMAP")
clustering <- data$louvain_SNNk5
counts <- as.matrix( counts(data)[ top.hvgs , ] )
```



### Basic processing with Seurat pipeline
***

If you have been using the `Seurat` pipeline so far. Below you can find a summary code for data processing until getting:

* A matrix where to perform the trajectory (UMAP)
* The cell clustering information

```{r, fig.height=10,fig.width=10}
suppressPackageStartupMessages({
library(Seurat)
library(cowplot)
})

#Data analysis with Seurat pipeline
data <- CreateSeuratObject(counts = umi_counts)
data <- NormalizeData(data)
data <- FindVariableFeatures(data, nfeatures = 2000)
data <- ScaleData(data)
data <- RunPCA(data)
data <- FindNeighbors(data)
data <- FindClusters(data,resolution = 1)
data <- RunUMAP(data, n.neighbors = 10, dims = 1:50,spread = 2,min.dist = 0.3 )

#Plot the clusters
DimPlot(data, group.by = "RNA_snn_res.1")

#Save the objects as separate matrices for input in slingshot
dimred <- data@reductions$umap@cell.embeddings
clustering <- data$RNA_snn_res.1
counts <- as.matrix( data@assays$RNA@counts[ data@assays$RNA@var.features , ] )
```


#### Trajectory inference with Slingshot
***

Until up to this point, the steps above have been covered in the previous lectures. From now on, we will start using that clustering and data reduction techniques for trajectory inference. The whole process can be done using a single function named `slingshot`, which is simply a wrapper for the 2 main steps for trajectory inference. The first step of the process is to define the lineages and then fit a curve through the data that defines a trajectory. These steps are break donw below for clarity.



#### Defining cell lineages with Slingshot

```{r}
suppressPackageStartupMessages({
library(slingshot)})

#Run default Slingshot lineage identification
set.seed(1)
lineages <- getLineages(data = dimred,
                     clusterLabels = clustering)

lineages

#Plot the lineages
par(mfrow=c(1,2))
plot(dimred[,1:2], col = pal[clustering],  cex=.5,pch = 16)
for(i in levels(clustering)){ 
text( mean(dimred[clustering==i,1]),
     mean(dimred[clustering==i,2]), labels = i,font = 2) }
plot(dimred[,1:2], col = pal[clustering],cex=.5, pch = 16)
lines(lineages, lwd = 3, col = 'black')
``` 


Here we see one central issue with trajectory analysis: where does the trajectory begin? Without any extra information, this is nearly an impossible task for a TI method. We need prior biological information to be able to define where the trajectory starts and where it should end.

```{r}
#Run default Slingshot
set.seed(1)
lineages <- getLineages(data = dimred,
                     clusterLabels = clustering,
                     #end.clus = c("11","7","10","9","5"), #define how many branches/lineages to consider
                     start.clus = "0") #define where to start the trajectories

lineages

#Plot the lineages
par(mfrow=c(1,2))
plot(dimred[,1:2], col = pal[clustering],  cex=.5,pch = 16)
for(i in levels(clustering)){ 
text( mean(dimred[clustering==i,1]),
     mean(dimred[clustering==i,2]), labels = i,font = 2) }
plot(dimred, col = pal[clustering],  pch = 16)
lines(lineages, lwd = 3, col = 'black')
```



#### Defining Principal Curves

Once the clusters are connected, Slingshot allows you to transform them to a smooth trajectory using principal curves. This is an algorithm that iteratively changes an initial curve to better match the data points. It was developed for linear data. To apply it to single-cell data, slingshot adds two enhancements:

* It will run principal curves for each ‘lineage’, which is a set of clusters that go from a defined start cluster to some end cluster
* Lineages with a same set of clusters will be constrained so that their principal curves remain bundled around the overlapping clusters

Since the function `getCurves()` takes some time to run, we can speed up the convergence of the curve fitting process by reducing the amount of cells to use in each lineage. Ideally you could all cells, but here we had set `approx_points` to 300 to speed up. Feel free to adjust that for your dataset.

```{r}
curves <- getCurves(lineages, approx_points = 300, thresh = 0.01, stretch = .8, allow.breaks = FALSE, shrink=.99)
curves

plot(dimred, col = pal[clustering], asp = 1, pch = 16)
lines(curves, lwd = 3, col = 'black')
```




### Finding differentially expressed genes
***

The main way to interpret a trajectory is to find genes that change along the trajectory. There are many ways to define differential expression along a trajectory:

* Expression changes along a particular path (i.e. change with pseudotime)
* Expression differences between branches
* Expression changes at branch points
* Expression changes somewhere along the trajectory
* …

`tradeSeq` is a recently proposed algorithm to find trajectory differentially expressed genes. It works by smoothing the gene expression along the trajectory by fitting a smoother using generalized additive models (GAMs), and testing whether certain coefficients are statstically different between points in the trajectory.

```{r}
BiocParallel::register(BiocParallel::SerialParam())
```

The fitting of GAMs can take quite a while, so for demonstration purposes we first do a very stringent filtering of the genes. In an ideal experiment, you would use all the genes, or at least those defined as being variable.

```{r}
library(tradeSeq)

#Removing some genes to speed up the computations for this tutorial
filt_counts <- counts [ rowSums(counts > 5) > ncol(counts)/100, ] 
dim(filt_counts)

sce <- fitGAM(  counts = as.matrix(filt_counts),
             sds = curves )

plotGeneCount(curves, filt_counts, clusters = clustering, models = sce)
```

```{r}
#Define function to plot
library(dplyr)
plot_differential_expression <- function(feature_id) {
feature_id <- pseudotime_association %>% filter(pvalue < 0.05) %>% top_n(1, -waldStat) %>% pull(feature_id)
cowplot::plot_grid(
plotGeneCount(curves, filt_counts, gene=feature_id[1], clusters = clustering, models = sce)+ ggplot2::theme(legend.position = "none"),
plotSmoothers(sce, as.matrix(counts), gene = feature_id[1])
)}

```



#### Genes that change with pseudotime

```{r}
pseudotime_association <- associationTest(sce)
pseudotime_association$fdr <- p.adjust(pseudotime_association$pvalue, method = "fdr")
pseudotime_association <- pseudotime_association[ order(pseudotime_association$pvalue), ]
pseudotime_association$feature_id <- rownames(pseudotime_association)
```


```{r}
feature_id <- pseudotime_association %>%
filter(pvalue < 0.05) %>%
top_n(1, -waldStat) %>%
pull(feature_id)
plot_differential_expression(feature_id)
```



#### Genes that change between two pseudotime points

We can define custom pseudotime values of interest if we’re interested in genes that change between particular point in pseudotime. By default, we can look at differences between start and end:

```{r}
pseudotime_start_end_association <- startVsEndTest(sce, pseudotimeValues = c(0, 1))
pseudotime_start_end_association$feature_id <- rownames(pseudotime_start_end_association)

feature_id <- pseudotime_start_end_association %>% 
filter(pvalue < 0.05) %>% 
top_n(1, waldStat) %>% 
pull(feature_id)

plot_differential_expression(feature_id)
```



#### Genes that are different between lineages

More interesting are genes that are different between two branches. We may have seen some of these genes already pop up in previous analyses of pseudotime. There are several ways to define “different between branches”, and each have their own functions:

* Different at the end points, using `diffEndTest`
* Different at the branching point, using `earlyDETest`
* Different somewhere in pseudotime the branching point, using `patternTest`
* Note that the last function requires that the pseudotimes between two lineages are aligned.

```{r}
different_end_association <- diffEndTest(sce)
different_end_association$feature_id <- rownames(different_end_association)
feature_id <- different_end_association %>% 
filter(pvalue < 0.05) %>% 
arrange(desc(waldStat)) %>% 
dplyr::slice(1) %>% 
pull(feature_id)
plot_differential_expression(feature_id)
```


```{r}
branch_point_association <- earlyDETest(sce)
branch_point_association$feature_id <- rownames(branch_point_association)

feature_id <- branch_point_association %>% 
filter(pvalue < 0.05) %>% 
arrange(desc(waldStat)) %>% 
dplyr::slice(1) %>% 
pull(feature_id)
plot_differential_expression(feature_id)
```



Check out this [vignette](https://statomics.github.io/tradeSeq/articles/tradeSeq.html) for a more in-depth overview of tradeSeq



### References
***

Cannoodt, Robrecht, Wouter Saelens, and Yvan Saeys. 2016. “Computational Methods for Trajectory Inference from Single-Cell Transcriptomics.” *European Journal of Immunology* 46 (11): 2496–2506. [doi](https://doi.org/10.1002/eji.201646347).

Saelens, Wouter, Robrecht Cannoodt, Helena Todorov, and Yvan Saeys. 2019. “A Comparison of Single-Cell Trajectory Inference Methods.” *Nature Biotechnology* 37 (5): 547–54. [doi](https://doi.org/10.1038/s41587-019-0071-9).

